### PR TITLE
Added ability to use custom partition keys for sinks

### PR DESCRIPTION
**Pull request by ashwinGokhale**

### CODE REVIEW

**Comment by snowplowcla** [comment]
Thanks for your pull request. Is this your first contribution to a Snowplow open source project? Before we can look at your pull request, you'll need to sign a Contributor License Agreement (CLA).

📝 Please visit https://github.com/snowplow/snowplow/wiki/CLA to learn more and sign.

Once you've signed, please reply here (e.g. I signed it!) and we'll verify. Thanks.

**Comment by ashwinGokhale** [comment]
I signed it!

**Comment by snowplowcla** [comment]
Confirmed! @ashwinGokhale has signed the Software Grant and Corporate Contributor License Agreement. Thanks so much.

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/AbstractSource.scala [hunk 1]
```diff
@@ -191,7 +191,16 @@ abstract class AbstractSource(config: KinesisEnrichConfig, igluResolver: Resolve
       validatedMaybeEvent match {
         case Success(co) => (tabSeparateEnrichedEvent(co), if (config.useIpAddressAsPartitionKey) {
             co.user_ipaddress
-          } else {
+          } 
+	  else if (config.partitionKey != null) {
+          try
+            String.valueOf(co.getClass.getDeclaredField(config.partitionKey).get(co))
+          catch {
+            case _: Exception =>
```
Comment (chuwy) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/AbstractSource.scala]: We never want to catch all exceptions. This should be something *similar* to 

```scala
  case _ @ (_: NoSuchFieldException | _: IllegalAccessException) => UUID.randomUUID.toString
```

But it is also very important to distinguish between cases where user tries to use non-existing property (must fail config validation long before runtime reaches this code) and when he tries to use `null` property.

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/AbstractSource.scala [hunk 2]
```diff
@@ -191,7 +191,16 @@ abstract class AbstractSource(config: KinesisEnrichConfig, igluResolver: Resolve
       validatedMaybeEvent match {
         case Success(co) => (tabSeparateEnrichedEvent(co), if (config.useIpAddressAsPartitionKey) {
             co.user_ipaddress
-          } else {
+          } 
+	  else if (config.partitionKey != null) {
+          try
+            String.valueOf(co.getClass.getDeclaredField(config.partitionKey).get(co))
```
Comment (chuwy) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/AbstractSource.scala]: I'm not entirely sure it is going to work. As all JVM fields in Scala are private. Access to them is made only through getters.

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/AbstractSource.scala [hunk 3]
```diff
@@ -191,7 +191,16 @@ abstract class AbstractSource(config: KinesisEnrichConfig, igluResolver: Resolve
       validatedMaybeEvent match {
         case Success(co) => (tabSeparateEnrichedEvent(co), if (config.useIpAddressAsPartitionKey) {
```
Comment (chuwy) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/AbstractSource.scala]: This whole `if-else` expression can be extracted as separate function with tests.

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/KinesisEnrichApp.scala [hunk 4]
```diff
@@ -378,4 +378,5 @@ class KinesisEnrichConfig(config: Config) {
   val maxBackoff = backoffPolicy.getLong("maxBackoff")
 
   val useIpAddressAsPartitionKey = outStreams.hasPath("useIpAddressAsPartitionKey") && outStreams.getBoolean("useIpAddressAsPartitionKey")
+  val partitionKey = if (outStreams.hasPath("partitionKey")) outStreams.getString("partitionKey") else null
```
Comment (chuwy) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/KinesisEnrichApp.scala]: Why not `Option[String]`?

File: 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/AbstractSource.scala (hunk 1 above)
Comment (chuwy) [inline 3-enrich/stream-enrich/src/main/scala/com.snowplowanalytics.snowplow.enrich/kinesis/sources/AbstractSource.scala]: Would also like to see some tests if this is going to be extracted into separate function.

**Comment by BenFradet** [comment]
Closing since this is a dupe of #3362 

@ashwinGokhale please try to keep only one PR open for the same issue.

**Pull request closed without merging**